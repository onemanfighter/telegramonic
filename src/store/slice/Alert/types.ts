export type AlertComponentState = {
  title: string;
  description: string;
  showIcon?: boolean;
  status?: 'error' | 'success' | 'warning' | 'info' | 'loading' | undefined;
  variant?: 'solid' | 'subtle' | 'left-accent' | 'top-accent';
};

export type AlertState = {
  alertVisibleState: boolean;
  alert: AlertComponentState;
};

export interface AlertStateAction {
  setAlertVisible: (visible: boolean) => void;
  setAlertStateData: (data: AlertComponentState) => void;
}

export type AlertStateSlice = AlertState & AlertStateAction;
