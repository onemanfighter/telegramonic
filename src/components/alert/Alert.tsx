import {
  Alert,
  AlertIcon,
  AlertTitle,
  AlertDescription,
} from '@chakra-ui/react';
import { AlertComponentProps } from './types';

const AlertComponent = ({
  title,
  description,
  status = 'info',
  variant = 'subtle',
  showIcon = true,
}: AlertComponentProps) => (
  <Alert status={status} variant={variant}>
    {showIcon && <AlertIcon />}
    <AlertTitle>{title}</AlertTitle>
    <AlertDescription>{description}</AlertDescription>
  </Alert>
);

export default AlertComponent;
