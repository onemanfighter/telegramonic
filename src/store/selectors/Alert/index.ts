export { AlertSelector } from './Alert.selector';
