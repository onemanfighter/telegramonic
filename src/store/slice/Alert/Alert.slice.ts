import { AppStoreSlice } from '../../appStore/types';
import { AlertState, AlertStateSlice } from './types';

const initialAlertState: AlertState = {
  alertVisibleState: false,
  alert: {
    title: '',
    description: '',
    showIcon: true,
    status: undefined,
    variant: 'subtle',
  },
};

const createAlertSlice: AppStoreSlice<AlertStateSlice> = (set) => ({
  ...initialAlertState,
  setAlertVisible: (visible) => {
    set((state) => {
      state.Alert.alertVisibleState = visible;
    });
  },
  setAlertStateData: (data) => {
    set((state) => {
      state.Alert.alert = data;
    });
  },
});

export default createAlertSlice;
