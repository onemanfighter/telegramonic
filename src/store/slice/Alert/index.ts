export { default as createAlertSlice } from './Alert.slice';
export type { AlertState, AlertComponentState, AlertStateSlice } from './types';
