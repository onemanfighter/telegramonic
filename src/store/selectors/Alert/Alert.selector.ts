import { AlertComponentState, AlertState } from '@store/slice';
import { AppStoreState } from 'src/store/appStore';

const AlertSelector = (state: AppStoreState) => ({
  visible: state.Alert.alertVisibleState,
  alert: state.Alert.alert,
  setAlertStateData: state.Alert.setAlertStateData,
  setAlertVisible: state.Alert.setAlertVisible,
  // Show alert for a specific time in seconds
  showAlertForTime: (data: AlertComponentState, time: number) => {
    state.Alert.setAlertStateData(data);
    state.Alert.setAlertVisible(true);
    setTimeout(() => {
      state.Alert.setAlertVisible(false);
    }, time * 1000);
  },
});

export { AlertSelector };
