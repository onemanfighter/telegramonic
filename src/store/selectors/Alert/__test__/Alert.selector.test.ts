import { AlertComponentState, appStore } from '@store';
import { act, renderHook } from '@testing-library/react';
import { AlertSelector } from '../Alert.selector';

jest.useFakeTimers();

describe('Alert selector', () => {
  const alertData: AlertComponentState = {
    description: 'description',
    showIcon: false,
    title: 'title',
    status: 'success',
    variant: 'solid',
  };
  it('should return alert selector', () => {
    const alertSelector = renderHook(() => appStore(AlertSelector)).result
      .current;

    expect(alertSelector).toMatchSnapshot();
  });

  it('should return alert state with initial value', () => {
    const alertSelector = renderHook(() => appStore(AlertSelector)).result
      .current;

    expect(alertSelector.alert).toEqual({
      description: '',
      showIcon: true,
      title: '',
      status: undefined,
      variant: 'subtle',
    });
  });

  it('should return alert state on alert data set using setAlertStateData', () => {
    const { result } = renderHook(() => appStore(AlertSelector));

    act(() => {
      result.current.setAlertStateData(alertData);
    });

    expect(result.current.alert).toMatchSnapshot();
  });

  it('setAlertVisible should update the visible state of the alert state', () => {
    const { result } = renderHook(() => appStore(AlertSelector));

    act(() => {
      result.current.setAlertVisible(true);
    });

    expect(result.current.visible).toBeTruthy();

    act(() => {
      result.current.setAlertVisible(false);
    });

    expect(result.current.visible).toBeFalsy();
  });

  it('showAlertForTime should set alert data and show the modal for given time', () => {
    const { result } = renderHook(() => appStore(AlertSelector));

    act(() => {
      result.current.showAlertForTime(alertData, 1000);
    });

    expect(result.current.alert).toEqual(alertData);
    expect(result.current.visible).toBeTruthy();

    act(() => {
      jest.runAllTimers();
    });

    expect(result.current.visible).toBeFalsy();
  });
});
