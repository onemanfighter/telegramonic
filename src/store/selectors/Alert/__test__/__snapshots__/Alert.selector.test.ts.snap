// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert selector should return alert selector 1`] = `
{
  "alert": {
    "description": "",
    "showIcon": true,
    "status": undefined,
    "title": "",
    "variant": "subtle",
  },
  "setAlertStateData": [Function],
  "setAlertVisible": [Function],
  "showAlertForTime": [Function],
  "visible": false,
}
`;

exports[`Alert selector should return alert state on alert data set using setAlertStateData 1`] = `
{
  "description": "description",
  "showIcon": false,
  "status": "success",
  "title": "title",
  "variant": "solid",
}
`;
